hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,749 @@
+//	var rg microsoftresourcesv1.ResourceGroup
+//	if err := gr.Client.Get(ctx, key, &rg); err != nil {
+//		if apierrors.IsNotFound(err) {
+//			// not able to find the resource, but that's ok. It might not exist yet
+//			return false, nil
+//		}
+//		return false, fmt.Errorf("error GETing rg resource with: %w", err)
+//	}
+//
+//	status, err := statusutil.GetProvisioningState(&rg)
+//	if err != nil {
+//		return false, fmt.Errorf("unable to get provisioning state for resource group with: %w", err)
+//	}
+//
+//	return zips.ProvisioningState(status) == zips.SucceededProvisioningState, nil
+//}
+
+func (gr *GenericReconciler) reconcileDelete(
+	ctx context.Context,
+	log logr.Logger,
+	metaObj genruntime.MetaObject,
+	f func(context.Context, logr.Logger, genruntime.ArmResource, genruntime.MetaObject) (ctrl.Result, error)) (ctrl.Result, error) {
+
+	log.Info("reconcile delete start")
+
+	_, armResourceSpec, err := ResourceSpecToArmResourceSpec(gr, metaObj)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "couldn't convert to armResourceSpec")
+	}
+	// TODO: Not setting status here
+	resource := genruntime.NewArmResource(armResourceSpec, nil, getResourceId(metaObj))
+
+	return f(ctx, log, resource, metaObj)
+}
+
+// startDeleteOfResource will begin the delete of a resource by telling Azure to start deleting it. The resource will be
+// marked with the provisioning state of "Deleting".
+func (gr *GenericReconciler) startDeleteOfResource(
+	ctx context.Context,
+	log logr.Logger,
+	resource genruntime.ArmResource,
+	metaObj genruntime.MetaObject) (ctrl.Result, error) {
+
+	if err := patcher(ctx, gr.Client, metaObj, func(ctx context.Context, mutMetaObject genruntime.MetaObject) error {
+		if resource.GetId() != "" {
+			emptyStatus, err := NewEmptyArmResourceStatus(metaObj)
+			if err != nil {
+				return errors.Wrapf(err, "failed trying to create empty status for %s", resource.GetId())
+			}
+
+			err = gr.ARMClient.BeginDeleteResource(ctx, resource.GetId(), resource.Spec().GetApiVersion(), emptyStatus)
+			if err != nil {
+				return errors.Wrapf(err, "failed trying to delete resource %s", resource.Spec().GetType())
+			}
+
+			annotations := make(map[string]string)
+			annotations[ResourceStateAnnotation] = string(armclient.DeletingProvisioningState)
+			addAnnotations(metaObj, annotations)
+		} else {
+			controllerutil.RemoveFinalizer(mutMetaObject, GenericControllerFinalizer)
+		}
+
+		// TODO: Not exactly sure what this was doing
+		//if err := gr.Converter.FromResource(resource, mutMetaObject); err != nil {
+		//	return errors.Wrapf(err, "error gr.Converter.FromResource")
+		//}
+
+		return nil
+	}); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, errors.Wrap(err, "failed to patch after starting delete")
+	}
+
+	// delete has started, check back to seen when the finalizer can be removed
+	return ctrl.Result{
+		RequeueAfter: 5 * time.Second,
```

Comment:
Are these timeouts used exactly, or does the main loop apply some windage?

If exact, we should modify the main loop to add a random offset (say, +/- 150ms) to avoid things getting into lockstep.

Reply:
I've taken a look at a few places where controllers have been written (ASO, etc) and looked through the controller-runtime documentation. I don't see examples where anybody is introducing jitter of their own and I don't see a location in the code where jitter is being introduced. @devigned wasn't sure either.

I've asked about it in the Kubernetes Slack and will see what they say.

Reply:
Here's the conversation I had:
> coderanger  27 minutes ago
There's some debounce in the reconcile queue I think, but otherwise I think it's handled directly.
coderanger  27 minutes ago
By default there is only one reconcile at a time (per controller) so load is smoothed somewhat naturally by that
matthchr  22 minutes ago
Is manually adding some jitter something you would recommend as a best practice?
I ask just because it's a common paradigm I've seen in other systems that are doing similar things to what a controller seems to do.
coderanger  12 minutes ago
No

Reply:
Rather than time we could have a "last state change" or something and do a time-based diff.

So if it's been 5m and we still haven't finished, wait another 1m, if it's been 30m and we still haven't deleted, wait another 5m type thing.

Reply:
I've made a note but won't do this for now - it can be done in a scoped smaller improvement where we introduce infrastructure for it.